sdk/include/platform/sunburst/platform-spi.hh

```diff
@@ -0,0 +1,403 @@
+		/**
+		 * Internal loopback function enabled when set to 1.
+		 */
+		ControlInternalLoopback = 1 << 30,
+		/**
+		 * Software reset performed when written as 1.
+		 */
+		ControlSoftwareReset = 1u << 31,
+	};
+
+	/// Status Register Fields
+	enum : uint32_t
+	{
+		/// Number of items in the transmit FIFO.
+		StatusTxFifoLevel = 0xffu << 0,
+		/// Number of items in the receive FIFO.
+		StatusRxFifoLevel = 0xffu << 8,
+		/**
+		 * When set the transmit FIFO is full and any data written to it will
+		 * be ignored.
+		 */
+		StatusTxFifoFull = 1u << 16,
+		/**
+		 * When set the receive FIFO is empty and any data read from it will be
+		 * undefined.
+		 */
+		StatusRxFifoEmpty = 1u << 17,
+		/// When set the SPI block is idle and can accept a new start command.
+		StatusIdle = 1u << 18,
+	};
+
+	/// Start Register Fields
+	enum : uint32_t
+	{
+		/// Number of bytes to receive/transmit in the SPI operation
+		StartByteCountMask = 0x7ffu,
+	};
+
+	/// Info Register Fields
+	enum : uint32_t
+	{
+		/// Maximum number of items in the transmit FIFO.
+		InfoTxFifoDepth = 0xffu << 0,
+		/// Maximum number of items in the receive FIFO.
+		InfoRxFifoDepth = 0xffu << 8,
+	};
+
+	/**
+	 * A driver for the Sonata's SPI device block.
+	 *
+	 * Documentation source can be found at:
+	 * https://github.com/lowRISC/sonata-system/blob/1a59633d2515d4fe186a07d53e49ff95c18d9bbf/doc/ip/spi.md
+	 *
+	 * Rendered documentation is served from:
+	 * https://lowrisc.org/sonata-system/doc/ip/spi.html
+	 */
+	template<size_t NumChipSelects = 4>
+	struct Generic : private utils::NoCopyNoMove
+	{
+		/**
+		 * The Sonata SPI block doesn't currently have support for interrupts.
+		 * The following registers are reserved for future use.
+		 */
+		uint32_t interruptState;
+		uint32_t interruptEnable;
+		uint32_t interruptTest;
+		/**
+		 * Configuration register. Controls how the SPI block transmits and
+		 * receives data. This register can be modified only whilst the SPI
+		 * block is idle.
+		 */
+		uint32_t configuration;
+		/**
+		 * Controls the operation of the SPI block. This register can
+		 * be modified only whilst the SPI block is idle.
+		 */
+		uint32_t control;
+		/// Status information about the SPI block
+		uint32_t status;
+		/**
+		 * Writes to this begin an SPI operation.
+		 * Writes are ignored when the SPI block is active.
+		 */
+		uint32_t start;
+		/**
+		 * Data from the receive FIFO. When read the data is popped from the
+		 * FIFO. If the FIFO is empty data read is undefined.
+		 */
+		uint32_t receiveFifo;
+		/**
+		 * Bytes written here are pushed to the transmit FIFO. If the FIFO is
+		 * full writes are ignored.
+		 */
+		uint32_t transmitFifo;
+		/**
+		 * Information about the SPI controller. This register reports the
+		 * depths of the transmit and receive FIFOs within the controller.
+		 */
+		uint32_t info;
+		/**
+		 * Chip Select lines; clear a bit to transmit/receive to/from the
+		 * corresponding peripheral.
+		 */
```

Comment:
What happens if this has a population count other than 1?

Reply:
All chip selects can be controlled individually, so one can have a population count of 0 (all enabled) to the number of chip selects (all disabled). (Chip selects are active low.)

https://github.com/lowRISC/sonata-system/blob/736837cda6de2d7b6dbbe86a513d18d2eb4155e2/rtl/ip/spi/rtl/spi.sv#L220-L226

Improved the comment to  better cover this.

Reply:
I apologize for the polarity error, but that wasn't really the question... SPI isn't (usually) an open-drain bus, and while SCK and MOSI might be OK with multiple receivers selected (and so, presumably, not tristated)... having multiple chip-selects active at once sure seems like it's going to do bad things to the drivers on MISO.

I know SPI isn't always bidirectional, and so maybe there are as few as zero drivers on MISO, but this really feels like something needs more blinking light warnings.